### PR TITLE
Training hooks

### PR DESCRIPTION
**Pull request by joejuzl**
**Proposed changes**:
Training hooks for caching and fingerprinting.

Still need to unit test, but was going to wait until we potentially have interface for caches and model storage.

### CODE REVIEW

File: rasa/engine/graph.py [hunk 1]
```diff
@@ -247,8 +295,36 @@ def __call__(self, *inputs_from_previous_nodes: Dict[Text, Any]) -> Dict[Text, A
...
+        for hook in self._hooks:
+            try:
+                hook_output = hook.on_before_node(
+                    node_name=self._node_name,
+                    config=self._component_config,
+                    received_inputs=received_inputs,
+                )
+                input_hook_outputs.append(hook_output)
+            except Exception as e:
+                raise GraphComponentException("Error running before hook.") from e
```
Comment (wochinge) [inline rasa/engine/graph.py:325]: how about also logging the node name?

File: rasa/engine/graph.py [hunk 2]
```diff
@@ -247,8 +295,36 @@ def __call__(self, *inputs_from_previous_nodes: Dict[Text, Any]) -> Dict[Text, A
...
+        for hook, hook_data in zip(self._hooks, input_hook_outputs):
+            try:
+                hook.on_after_node(
+                    node_name=self._node_name,
+                    config=self._component_config,
+                    output=output,
+                    input_hook_data=hook_data,
+                )
+            except Exception as e:
+                raise GraphComponentException("Error running after hook.") from e
```
Comment (wochinge) [inline rasa/engine/graph.py:312]: how about logging the node name?

File: rasa/engine/training/hooks.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+class TrainingHook(GraphNodeHook):
+    """Caches fingerprints and outputs of nodes during model training."""
+
+    def __init__(self, cache: TrainingCache, model_storage: ModelStorage):
+        """Initializes a `TrainingHook`.
+
+        Args:
+            cache: Cache used to store fingerprints and outputs
+            model_storage: Used to cache `Resource`s
```
Comment (wochinge) [inline rasa/engine/training/hooks.py:17]:
```suggestion
            cache: Cache used to store fingerprints and outputs.
            model_storage: Used to cache `Resource`s.
```

File: rasa/engine/training/hooks.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        self._model_storage = model_storage
+
+    def on_before_node(
+        self,
+        node_name: Text,
+        config: Dict[Text, Any],
+        received_inputs: Dict[Text, Any],
+    ) -> Dict:
+        """Calculates the run fingerprint for use in `on_after_node`."""
+        fingerprint_key = rasa.shared.utils.io.deep_container_fingerprint(
```
Comment (wochinge) [inline rasa/engine/training/hooks.py:29]: should we add a debug log in `on_before_node` and `on_after_node`?

File: tests/engine/graph_components_test_classes.py [hunk 5]
```diff
@@ -176,3 +178,39 @@ def run_train_process(self) -> Any:
 
     def run_inference(self) -> Any:
         return self._eager_instantiated_value
+
+
+class CacheableText:
+    def __init__(self, text: Text):
```
Comment (wochinge) [inline tests/engine/graph_components_test_classes.py:184]:
```suggestion
    def __init__(self, text: Text) -> None:
```

File: tests/engine/training/test_hooks.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+from rasa.engine.model_storage import ModelStorage
+from rasa.engine.training.hooks import TrainingHook
+import rasa.shared.utils.io
+from tests.engine.graph_components_test_classes import (
+    CacheableComponent,
+    CacheableText,
+)
+
+
+def test_training_hook_saves_to_cache(
```
Comment (wochinge) [inline tests/engine/training/test_hooks.py:12]: we validate that everything is fingerprintable beforehand right? Do we need to handle the case that something is not fingerprintable although the type annotations suggested that something is fingerprintable?

Comment (joejuzl) [reply tests/engine/training/test_hooks.py:12]: I.e. an error within the `.fingerprint` method?

Comment (wochinge) [reply tests/engine/training/test_hooks.py:12]: yeah, maybe in the `deep_container` one or in the extracted one which we just spoke about

File: tests/engine/graph_components_test_classes.py [hunk 7]
```diff
@@ -177,3 +178,39 @@ def run_train_process(self) -> Any:
...
+        cls, node_name: Text, directory: Path, model_storage: ModelStorage
+    ) -> CacheableText:
+        text = rasa.shared.utils.io.read_file(directory / "my_file.txt")
+        return cls(text=text)
+
+
+class CacheableComponent(GraphComponent):
+    default_config = {"prefix": "Hello "}
+
+    def __init__(self, prefix: Text):
```
Comment (wochinge) [inline tests/engine/graph_components_test_classes.py:201]:
```suggestion
    def __init__(self, prefix: Text) -> None:
```

Comment (wochinge) [reply tests/engine/graph_components_test_classes.py:201]: optional: Could make that a dataclass to get rid of the constructor?

File: rasa/engine/graph.py [hunk 8]
```diff
@@ -159,6 +159,47 @@ def required_packages(self) -> List[Text]:
...
+        output: Any,
+        input_hook_data: Dict,
+    ) -> None:
+        """Runs after the `GraphNode` as executed.
+
+        Args:
+            node_name: The name of the node that has run.
+            config: The node's config.
+            output: The output of the node.
+            input_hook_data: Data returned rom `on_before_node`.
```
Comment (wochinge) [inline rasa/engine/graph.py:198]:
```suggestion
            input_hook_data: Data returned from `on_before_node`.
```

**Review by wochinge — APPROVED** [review]

**Comment by wochinge** [comment]
💯

**Pull request merged**